### PR TITLE
Gospel Statistics

### PR DESCRIPTION
**Pull request by zioninzion**
Trainee side only

### CODE REVIEW

File: ap/gospel_statistics/models.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  # Which trainees composing this gospel pair
+  team = models.ForeignKey(Team, blank=True, null=True, on_delete=models.SET_NULL)
+  term = models.ForeignKey(Term, blank=True, null=True, on_delete=models.SET_NULL)
+  trainees = models.ManyToManyField(Trainee, related_name="gospel_statistics")
+
+
+# Gospel Statistics
+class GospelStat(models.Model):
+  # Might want to change SET_NULL to CASCADE
+  gospelpair = models.ForeignKey(GospelPair, blank=True, null=True, on_delete=models.SET_NULL)
```
Comment (iamsitting) [inline ap/gospel_statistics/models.py]: Yes, let's do cascade.

File: ap/gospel_statistics/models.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+from terms.models import Term
+
+
+# For which gospel pair there should be exactly one gospel statistics for every week
+# Gospel Pair
+class GospelPair(models.Model):
+  # Which trainees composing this gospel pair
+  team = models.ForeignKey(Team, blank=True, null=True, on_delete=models.SET_NULL)
+  term = models.ForeignKey(Term, blank=True, null=True, on_delete=models.SET_NULL)
+  trainees = models.ManyToManyField(Trainee, related_name="gospel_statistics")
```
Comment (iamsitting) [inline ap/gospel_statistics/models.py]: Write a __unicode__ method for this model.

File: ap/gospel_statistics/views.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        entry[_attributes[12]] += stat.conference
+      data.append(entry)
+    return data
+
+  def post(self, request, *args, **kwargs):
+    # Retreive the updated stat values
+    list_of_pairs = request.POST.getlist('pairs')
+    list_of_stats = request.POST.getlist('inputs')
+    current_week = get_week()
+    for i in range(len(list_of_pairs)):
```
Comment (iamsitting) [inline ap/gospel_statistics/views.py]: Use enumerate instead of range-len

File: ap/gospel_statistics/views.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  template_name = "gospel_statistics/gospel_statistics.html"
+
+  @staticmethod
+  def get_stats_list(gospel_pairs, gospel_statistics):
+    data = []
+    for p in gospel_pairs:
+      entry = dict()
+      entry['gospel_pair'] = p
+      # stat = gospel_statistics.filter(gospelpair=p, week=get_week())[0]
+      for i in range(len(_attributes)):
```
Comment (iamsitting) [inline ap/gospel_statistics/views.py]: Use enumerate instead of range-len

File: ap/gospel_statistics/views.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    'bible_study', 'small_group', 'district_meeting', 'conference'
+]
+ctx = dict()
+for i in _attributes:
+  ctx[i] = 0
+
+
+def get_week():
+  for i in range(0, 21):
+    if Term.current_term().startdate_of_week(i) <= date.today() \
```
Comment (iamsitting) [inline ap/gospel_statistics/views.py]: You use Term.current_term() a lot. Try creating a global variable C_TERM = Term.current_term() and using that instead.

File: ap/gospel_statistics/views.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    context['all_stat'] = self.get_all_stats_list(context['gospel_pairs'], GospelStat.objects.filter(gospelpair__in=context['gospel_pairs']))
+    return context
+
+
+class NewGospelPairView(TemplateView):
+  template_name = "gospel_statistics/new_pair.html"
+
+  def post(self, request, *args, **kwargs):
+    # Do we need this?
+    context = self.get_context_data()
```
Comment (iamsitting) [inline ap/gospel_statistics/views.py]: You would need context data if you were using render, but instead you are using redirect. So, no.

**Comment by joysyu** [comment]
Looks good!
- When selecting "Week Of", it doesn't save the selected option in the selection box.
- The webpage just breaks when there's an error (e.g. user inputs text instead of numbers as data).

**Comment by Plew** [comment]
Moving this to #1615

**Pull request closed without merging**